opm/autodiff/ParallelDebugOutput.hpp

```diff
@@ -0,0 +1,553 @@
+                for( auto it = localWellState_.wellMap().begin(); it != end; ++it )
+                {
+                    const std::string& name = it->first;
+                    const int wellIdx = it->second[ 0 ];
+
+                    // write well name
+                    writeString( buffer, name );
+
+                    // write well data
+                    buffer.write( localWellState_.bhp()[ wellIdx ] );
+                    buffer.write( localWellState_.thp()[ wellIdx ] );
+                    const int wellRateIdx = wellIdx * localWellState_.numPhases();
+                    for( int np=0; np<localWellState_.numPhases(); ++np )
+                        buffer.write( localWellState_.wellRates()[ wellRateIdx + np ] );
+
+                    // TODO: perfRates and perfPress, need to figure out the index
+                    // mapping there.
+                }
+            }
+
+            void readWells( MessageBufferType& buffer )
+            {
+                int nWells = -1;
+                buffer.read( nWells );
+                // unpack all wells that have been sent
+                std::string name ;
+                for( int well = 0; well < nWells ; ++well )
+                {
+                    // read well name for local identification
+                    readString( buffer, name );
+
+                    // unpack values
+                    auto it = globalWellState_.wellMap().find( name );
+                    if( it == globalWellState_.wellMap().end() )
+                    {
+                        OPM_THROW(std::logic_error,"global state does not contain well " <<  name );
+                    }
+                    const int wellIdx = it->second[ 0 ];
+
+                    buffer.read( globalWellState_.bhp()[ wellIdx ] );
+                    buffer.read( globalWellState_.thp()[ wellIdx ] );
+                    const int wellRateIdx = wellIdx * globalWellState_.numPhases();
+                    for( int np=0; np<globalWellState_.numPhases(); ++np )
+                        buffer.read( globalWellState_.wellRates()[ wellRateIdx + np ] );
+
+                    // TODO: perfRates and perfPress, need to figure out the index
+                    // mapping there.
+                }
+            }
+
+        };
+
+        // gather solution to rank 0 for EclipseWriter
+        bool collectToIORank( const SimulatorState& localReservoirState,
+                              const WellState& localWellState )
+        {
+            PackUnPackSimulatorState packUnpack( localReservoirState, globalReservoirState_,
+                                                 localWellState, globalWellState_,
+                                                 localIndexMap_, indexMaps_, isIORank() );
+
+            //toIORankComm_.exchangeCached( packUnpack );
+            toIORankComm_.exchange( packUnpack );
+#ifndef NDEBUG
+            // mkae sure every process is on the same page
+            toIORankComm_.barrier();
```

Comment:
Good old gdb with a simple script:
mpirun -np 2 run_gdb.sh <executable>
```sh
#!/bin/sh

echo "Running GDB on node `hostname`"
dir=`pwd`
if [ -e gdbfile ] ; then
        xterm -e "gdb -x $dir/gdbfile --args $* "
else
        xterm -e "gdb --args $* "
fi
exit 0
```
That fires up a xterm with a gdb session for each process. You can create a file gdbfile that contains commands for gdb (e.g. run comand, breakpoints, etc.)